### PR TITLE
MIFOSX-2149-Surplus Payment Treatment for Loan

### PR DESCRIPTION
**Pull request by ippezshelby**

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -1789,6 +1791,45 @@ private LoanCharge retrieveLoanChargeBy(final Long loanId, final Long loanCharge
...
+    @CronTarget(jobName = JobName.ALLOCATE_OVERPAYMENTS_TO_SAVINGS)
+    public void allocateOverpayments() throws JobExecutionException {
+        final Collection<Long> loanIds = this.loanReadPlatformService.fetchOverpayedLoansForAllocation();
+        final StringBuilder sb = new StringBuilder();
+
+        if(loanIds != null){
+            for(final Long loanId : loanIds){
+
+                try{
+                    final AccountAssociations accountAssociations = this.accountAssociationRepository.findByLoanIdAndType(loanId,AccountAssociationType.LINKED_ACCOUNT_ASSOCIATION.getValue());
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: What happens when linked savings account is not available? There might be a null pointer exception here.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/data/LoanProductData.java [hunk 2]
```diff
@@ -117,6 +117,7 @@
     private final Integer minimumDaysBetweenDisbursalAndFirstRepayment;
     private final boolean canDefineInstallmentAmount;
     private final Integer installmentAmountInMultiplesOf;
+	private final boolean canAutoAllocateOverpayments;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/data/LoanProductData.java:120]: I believe, when canAutoAllocateOverpayments is enabled from loan product definition, then we need to mandate linked savings account for the loans created from this product

**Review by nazeer1100126 — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 1 above)
Comment (ippezshelby) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I thought 
if(loanIds != null){ 
Will handle this exception

Comment (nazeer1100126) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Assume a case with over paid loans which doesn't have linked savings account. That time accountAssociations will be null right?

**Pull request closed without merging**